### ISSUE TITLE
Python bindings leak references

### ISSUE THREAD

**Issue by hedger** [unicorn-engine/unicorn#710]
Due to `Uc` class having `__del__` method and hooks/callbacks creating circular references to `Uc` instances, they are never garbage-collected and remain in memory. In scenarios where user needs to create many `Uc` instances, they are never GC'd and quickly take up all available memory.

Refs: https://docs.python.org/2/library/gc.html#gc.garbage
https://docs.python.org/2/reference/datamodel.html?highlight=__del__#object.__del__

In the following test case, a hook is added and removed immediately. But it remains in `_ctype_cbs` and prevents GC from collecting `Uc` instance.
```
#!/usr/bin/python

from __future__ import print_function

import time

from unicorn import *
from unicorn.x86_const import *

import regress

ADDRESS = 0x8048000
STACK_ADDRESS = 0xffff000
STACK_SIZE = 4096
'''
31 DB           xor ebx, ebx
53              push ebx
43              inc ebx
53              push ebx
6A 02           push 2
6A 66           push 66h
58              pop eax
89 E1           mov ecx, esp
CD 80           int 80h
'''
CODE = "\x31\xDB\x53\x43\x53\x6A\x02\x6A\x66\x58\x89\xE1\xCD\x80"
EP = ADDRESS + 0x54

def hook_code(mu, address, size, user_data):
        print(">>> Tracing instruction at 0x%x, instruction size = %u" %(address, size))
        
def emu_loop():
    emu = Uc(UC_ARCH_X86, UC_MODE_32)
    emu.mem_map(ADDRESS, 0x1000)
    emu.mem_write(EP, CODE)

    emu.mem_map(STACK_ADDRESS, STACK_SIZE)
    emu.reg_write(UC_X86_REG_ESP, STACK_ADDRESS + STACK_SIZE)
    
    i = emu.hook_add(UC_HOOK_CODE, hook_code, None)
    emu.hook_del(i)
    
    emu.emu_start(EP, EP + len(CODE), count = 3)
    print("EIP: 0x%x" % emu.reg_read(UC_X86_REG_EIP))

def debugMem():
    import gc
    gc.collect()  # don't care about stuff that would be garbage collected properly
    print("Orphaned objects in gc.garbage:", gc.garbage)
    assert(len(gc.garbage) == 0)

class EmuLoopReferenceTest(regress.RegressTest):
    def runTest(self):
        for i in range(5):
            emu_loop()
        debugMem()


if __name__ == '__main__':
    regress.main()
```

Test results on current version:
```
>leaked_refs.py
EIP: 0x8048058
EIP: 0x8048058
EIP: 0x8048058
EIP: 0x8048058
EIP: 0x8048058
Orphaned objects in gc.garbage: [<unicorn.unicorn.Uc object at 0x0364B0D0>, <unicorn.unicorn.Uc object at 0x0364B170>, <unicorn.unicorn.Uc object at 0x0364B1B0>, <unicorn.unicorn.Uc object at 0x0364B1D0>, <unicorn.unicorn.Uc object at 0x0364B1F0>]
F
======================================================================
FAIL: runTest (__main__.EmuLoopReferenceTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "H:\code\unicorn\tests\regress\leaked_refs.py", line 56, in runTest
    debugMem()
  File "H:\code\unicorn\tests\regress\leaked_refs.py", line 50, in debugMem
    assert(len(gc.garbage) == 0)
AssertionError

----------------------------------------------------------------------
Ran 1 test in 0.016s

FAILED (failures=1)
```

**Comment by aquynh** [comment]
can you fix this issue?

**Comment by hedger** [comment]
Fixed, #713.
